Firestore/core/test/firebase/firestore/model/field_path_test.cc

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/model/field_path.h"
+
+#include <initializer_list>
+#include <string>
+#include <vector>
+
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+TEST(FieldPath, Constructors) {
+  const FieldPath empty_path;
+  EXPECT_TRUE(empty_path.empty());
+  EXPECT_EQ(0, empty_path.size());
+  EXPECT_TRUE(empty_path.begin() == empty_path.end());
+
+  const FieldPath path_from_list = {"rooms", "Eros", "messages"};
+  EXPECT_FALSE(path_from_list.empty());
+  EXPECT_EQ(3, path_from_list.size());
+  EXPECT_TRUE(path_from_list.begin() + 3 == path_from_list.end());
+
+  std::vector<std::string> segments{"rooms", "Eros", "messages"};
+  const FieldPath path_from_segments{segments.begin(), segments.end()};
+  EXPECT_FALSE(path_from_segments.empty());
+  EXPECT_EQ(3, path_from_segments.size());
+  EXPECT_TRUE(path_from_segments.begin() + 3 == path_from_segments.end());
+
+  FieldPath copied = path_from_list;
+  EXPECT_EQ(path_from_list, copied);
+  const FieldPath moved = std::move(copied);
+  // Because FieldPath is immutable, move constructor performs a copy.
+  EXPECT_EQ(copied, moved);
+}
+
+TEST(FieldPath, Indexing) {
+  const FieldPath path{"rooms", "Eros", "messages"};
+
+  EXPECT_EQ(path.front(), "rooms");
```

Comment:
Do we really need 3 ways to say the same thing?

Reply:
No. I suggest removing `at`.

`front` and `[0]` are necessary because the analogous `back` and `path[path.size() - 1]` are not so easy.

Reply:
Removed `at`.